### PR TITLE
fix(ngAnimate): only buffer rAF requests within the animation runners

### LINKED ISSUE #12280: $$rAF performance issue in Angular Material
When scrolling *sticky* md-subheader, recent changes in $$rAF seem to create excessive CPU usages.
See https://github.com/angular/material/issues/3531 for details and CodePen demos.

### PR DESCRIPTION
**Pull request by matsko**
Closes #12280

Here are demos to show that this works (test in IE):

Multiple RAF Requests in Angular 1.4 ([#11791](https://github.com/angular/angular.js/issues/11791))
- Pre 1.4.0 functionality (really slow on **IE 11** ... this was the reason why we decided to bundle the calls into a single function): https://jsfiddle.net/oga2z7bb/3/
- Current functionality (fixed for **IE 11**): https://angular-development.appspot.com/raf-perf/fast/index.html
- Patched functionality (fixed for **IE 11**): https://angular-development.appspot.com/raf-perf/slow/index.html

Material Design $$rAF perf issue ([#12280](https://github.com/angular/angular.js/issues/12280))
- Current functionality (**slow**): https://angular-development.appspot.com/slow-raf/slow/index.html
- Patched functionality (**fast**): https://angular-development.appspot.com/slow-raf/fast/index.html

### CODE REVIEW

**Comment by matsko** [comment]
@ThomasBurleson @AdriVanHoudt @MichhDiego this should fix the issues you guys are having.

**Comment by matsko** [comment]
@IgorMinar please review.

File: test/ngRoute/directive/ngViewSpec.js [hunk 1]
```diff
@@ -913,9 +912,11 @@ describe('ngView animations', function() {
 
       $location.path('/foo');
       $rootScope.$digest();
-      expect($animate.queue.shift().event).toBe('enter');
-      $animate.triggerCallbacks();
 
+      $animate.flush();
+      $rootScope.$digest();
```
Comment (IgorMinar) [inline test/ngRoute/directive/ngViewSpec.js]: why do we need the extra digest after flush? it wasn't necessary before and we digest just before flushing already.

File: src/ngAnimate/animateRunner.js [hunk 2]
```diff
@@ -1,18 +1,33 @@
 'use strict';
 
-var $$rAFMutexFactory = ['$$rAF', function($$rAF) {
+var $$AnimateMutexFactory = ['$$rAF', function($$rAF) {
```
Comment (IgorMinar) [inline src/ngAnimate/animateRunner.js]: Why do you call this a mutex? it's not quite the same as regular mutex. How about $$AnimateAsyncRun or $$AnimateAsyncExecutor?

File: src/ngMock/angular-mocks.js [hunk 3]
```diff
@@ -763,25 +763,50 @@ angular.mock.animate = angular.module('ngAnimateMock', ['ng'])
...
-        triggerCallbackEvents: function() {
-          $$rAF.flush();
-        },
-        triggerCallbackPromise: function() {
-          $timeout.flush(0);
-        },
-        triggerCallbacks: function() {
-          this.triggerCallbackEvents();
-          this.triggerCallbackPromise();
+        flush: function() {
```
Comment (IgorMinar) [inline src/ngMock/angular-mocks.js]: didn't you want to document the flush method as public api?

**Comment by IgorMinar** [comment]
the rest looks good. thanks!

**Comment by AdriVanHoudt** [comment]
thank you so much :clap:

File: src/ngMock/angular-mocks.js (hunk 3 above)
Comment (matsko) [inline src/ngMock/angular-mocks.js]: Let's do this one in a follow up commit later this week. That way we can have a page in the guide that explains animation testing.

File: src/ngAnimate/animateRunner.js (hunk 2 above)
Comment (matsko) [inline src/ngAnimate/animateRunner.js]: Changed to `$$AnimateAsyncRun`

File: test/ngRoute/directive/ngViewSpec.js (hunk 1 above)
Comment (matsko) [inline test/ngRoute/directive/ngViewSpec.js]: So beforehand what happened was:

1. `triggerCallbacks()` issued a `$$rAF.flush()` internally when called.
2. `triggerCallbacks()` also calls `$timeout.flush(0)` which calls a digest somewhere down the line.

But with the changes to use `$animate.flush()` we end up having to do this:

1. `$animate.flush()` clears out the RAF for the runner.
2. `$rootScope.$digest()` allows for the resolved promise to flush (which can only be done after the RAF has flushed).

So we need to call RAF + digest always.

**Comment by matsko** [comment]
Closed with https://github.com/angular/angular.js/commit/dc48aadd26bbf1797c1c408f63ffde99d67414a9

**Pull request closed without merging**